### PR TITLE
Custom schema

### PR DESCRIPTION
**Pull request by pjoshi04**

### CODE REVIEW

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java [hunk 1]
```diff
@@ -61,10 +61,6 @@ public void registerSubject(long occurrences) {
     this.subjectsWithThisPredicate += occurrences;
   }
 
-  public boolean isOptional() {
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java]: Why is optional removed?

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java [hunk 2]
```diff
@@ -96,6 +102,11 @@ public boolean hasBeenList() {
     return hasBeenList;
   }
 
+  @JsonProperty("isExplicit")
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java]: Conflicts with the @JsonIgnore of isExplicit

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/ExplicitField.java [hunk 3]
```diff
@@ -23,7 +23,7 @@
   @JsonProperty("reference")
   private List<String> references;
   @JsonProperty("type")
-  private ExplicitType explicitType;
+  private ExplicitType explicitType; //todo:remove this
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/ExplicitField.java]: Why not remove them immediately

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/graphql/mutations/MergeSchemaDataFetcher.java [hunk 4]
```diff
@@ -84,12 +85,25 @@ public Object get(DataFetchingEnvironment env) {
     File customSchemaFile = fileHelper.fileInDataSet(ownerIdDataSetName.getLeft(), ownerIdDataSetName.getRight(),
       "customSchema.json");
 
+    Map<String, List<ExplicitField>> existingCustomSchema;
+
     try {
-      objectMapper.writeValue(customSchemaFile, customSchema);
+      existingCustomSchema = objectMapper.readValue(customSchemaFile,
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/graphql/mutations/MergeSchemaDataFetcher.java]: Add check that the file exists before trying to read the value. This prevents a lot of exceptions.

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/graphql/rootquery/RootQuery.java [hunk 5]
```diff
@@ -335,18 +330,21 @@ public synchronized GraphQLSchema rebuildSchema() {
         "customSchema.json");
 
       try {
-        customSchema = objectMapper.readValue(customSchemaFile, new TypeReference<List<ExplicitType>>(){});
+        customSchema = objectMapper.readValue(customSchemaFile,
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/graphql/rootquery/RootQuery.java]: Duplicate code with CustomSchemaDataFetcher.

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/implementations/bdb/MergeExplicitSchemas.java [hunk 6]
```diff
@@ -69,7 +67,7 @@ private ExplicitField findAndMergeExplicitFields(String collection, ExplicitFiel
     }
 
     if (explicitField2 != null && explicitField1.getUri().equals(explicitField2.getUri())) {
-      mergedField = mergeExplicitFields.mergeExplicitFields(explicitField1, explicitField2);
+      mergedField = explicitField1.mergeWith(explicitField2);
```
Comment (MMaas3) [inline timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/implementations/bdb/MergeExplicitSchemas.java]: Does this code add all the fields of explicitieschema2?

File: timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java (hunk 1 above)
Comment (pjoshi04) [reply timbuctoo-instancev4/src/main/java/nl/knaw/huygens/timbuctoo/v5/datastores/schemastore/dto/Predicate.java]: Ask Jauco why we did not need isOptional anymore here.

**Pull request closed without merging**